File: riistaandroid/src/main/java/fi/riista/mobile/pages/ShootingTestCalendarEventListFragment.java
package fi.riista.mobile.pages;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ListView;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;
import javax.inject.Named;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import dagger.android.support.AndroidSupportInjection;
import fi.riista.mobile.R;
import fi.riista.mobile.activity.ShootingTestMainActivity;
import fi.riista.mobile.adapter.ShootingTestCalendarEventsAdapter;
import fi.riista.mobile.models.shootingTest.ShootingTestCalendarEvent;
import fi.riista.mobile.network.shootingTest.ListShootingCalendarEventsTask;
import fi.vincit.androidutilslib.context.WorkContext;

import static fi.riista.mobile.di.DependencyQualifiers.APPLICATION_WORK_CONTEXT_NAME;

public class ShootingTestCalendarEventListFragment extends PageFragment {

    public static final String EXTRA_CALENDAR_EVENT_ID = "extra_calendar_event_id";

    @Inject
    @Named(APPLICATION_WORK_CONTEXT_NAME)
    WorkContext mAppWorkContext;

    private List<ShootingTestCalendarEvent> mDataModels = new ArrayList<>();
    private ShootingTestCalendarEventsAdapter mAdapter;

    public static ShootingTestCalendarEventListFragment newInstance() {
        final ShootingTestCalendarEventListFragment fragment = new ShootingTestCalendarEventListFragment();
        fragment.setArguments(new Bundle());
        return fragment;
    }

    // Dagger injection of a Fragment instance must be done in On-Attach lifecycle phase.
    @Override
    public void onAttach(@NonNull final Context context) {
        AndroidSupportInjection.inject(this);
        super.onAttach(context);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull final LayoutInflater inflater,
                             @Nullable final ViewGroup container,
                             @Nullable final Bundle savedInstanceState) {

        final View view = inflater.inflate(R.layout.fragment_shooting_test_list, container, false);

        setupActionBar(R.layout.actionbar_shooting_test_list, false);

        mAdapter = new ShootingTestCalendarEventsAdapter(mDataModels, getActivity());

        final ListView listView = view.findViewById(R.id.shooting_test_list);
        listView.setAdapter(mAdapter);
        listView.setOnItemClickListener((parent, view1, position, id) -> {
            final ShootingTestCalendarEvent clickedEvent = mDataModels.get(position);

            final Intent intent = new Intent(getActivity(), ShootingTestMainActivity.class);
            intent.putExtra(EXTRA_CALENDAR_EVENT_ID, clickedEvent.calendarEventId);
            startActivity(intent);
        });

        return view;
    }

    @Override
    public void onResume() {
        super.onResume();

        refreshData();
    }

    private void refreshData() {
        final ListShootingCalendarEventsTask task = new ListShootingCalendarEventsTask(mAppWorkContext) {
            @Override
            protected void onFinishObjects(final List<ShootingTestCalendarEvent> results) {
                mDataModels = results;

                mAdapter.clear();
                mAdapter.addAll(mDataModels);
                mAdapter.notifyDataSetChanged();
            }

            @Override
            protected void onError() {
                super.onError();
            }
        };
        task.start();
    }
}

File: riistaandroid/src/main/java/fi/riista/mobile/pages/ShootingTestCalendarEventListFragment.kt
package fi.riista.mobile.pages

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.AdapterView.OnItemClickListener
import android.widget.ListView
import fi.riista.common.RiistaSDK
import fi.riista.common.domain.OperationResultWithData
import fi.riista.mobile.R
import fi.riista.mobile.activity.ShootingTestMainActivity
import fi.riista.mobile.adapter.ShootingTestCalendarEventsAdapter
import fi.riista.mobile.models.shootingTest.ShootingTestCalendarEvent
import fi.riista.mobile.riistaSdkHelpers.toShootingTestCalendarEvent
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch

class ShootingTestCalendarEventListFragment : PageFragment() {
    private var dataModels: List<ShootingTestCalendarEvent> = mutableListOf()
    private lateinit var adapter: ShootingTestCalendarEventsAdapter
    private val shootingTestContext = RiistaSDK.shootingTestContext

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.fragment_shooting_test_list, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setupActionBar(R.layout.actionbar_shooting_test_list, false)
        adapter = ShootingTestCalendarEventsAdapter(dataModels, activity)
        val listView = view.findViewById<ListView>(R.id.shooting_test_list)
        listView.adapter = adapter
        listView.onItemClickListener =
            OnItemClickListener { _: AdapterView<*>?, _: View?, position: Int, _: Long ->
                val clickedEvent = dataModels[position]
                val intent = Intent(activity, ShootingTestMainActivity::class.java)
                intent.putExtra(EXTRA_CALENDAR_EVENT_ID, clickedEvent.calendarEventId)
                startActivity(intent)
            }
    }

    override fun onResume() {
        super.onResume()
        refreshData()
    }

    private fun refreshData() {
        MainScope().launch {
            val response = shootingTestContext.fetchShootingTestCalendarEvents()

            if (!isResumed) {
                return@launch
            }

            when (response) {
                is OperationResultWithData.Success -> {
                    dataModels = response.data.map { it.toShootingTestCalendarEvent() }
                    adapter.clear()
                    adapter.addAll(dataModels)
                    adapter.notifyDataSetChanged()
                }
                else -> {
                    // No-op
                }
            }
        }
    }

    companion object {
        const val EXTRA_CALENDAR_EVENT_ID = "extra_calendar_event_id"
    }
}
